elide-datastore/elide-datastore-aggregation/src/main/java/com/yahoo/elide/datastores/aggregation/QueryEngine.java

```diff
@@ -0,0 +1,161 @@
+     * <p>
+     * This method calls {@link #buildQuery(Class, Optional, Optional, Optional, Optional, RequestScope)}. If there is
+     * a filter associated with the query, the filter is splitted into a WHERE and a HAVING filter by calling
+     * {@link #extractWhereExpression(Optional)} and {@link #extractHavingExpression(Optional)}, respectively.
+     *
+     * @param entityClass  The type of the entity involved in this query
+     * @param filterExpression  Filters that can be evaluated in the data store. It is optional for the data store to
+     * attempt evaluation.
+     * @param sorting  Sorting which can be pushed down to the data store.
+     * @param pagination  Pagination which can be pushed down to the data store.
+     * @param scope  Contains request level metadata.
+     *
+     * @return a optimized query consumed by the data store
+     */
+    default Query buildQuery(
+            Class<?> entityClass,
+            Optional<FilterExpression> filterExpression,
+            Optional<Sorting> sorting,
+            Optional<Pagination> pagination,
+            RequestScope scope
+    ) {
+        return buildQuery(
+                entityClass,
+                extractWhereExpression(filterExpression),
+                extractHavingExpression(filterExpression),
+                sorting,
+                pagination,
+                scope
+        );
+    }
+
+    /**
+     * Parses request to generate a {@link Query} object to be executed by this {@link QueryEngine}.
+     * <p>
+     * This method should also incorporate all query optimizations if any
+     *
+     * @param entityClass  The type of the entity involved in this query
+     * @param whereFilter  A WHERE filter that can be evaluated in the data store. It is optional for the data store to
+     * attempt evaluation
+     * @param havingFilter  A HAVING filter that can be evaluated in the data store. It is optional for the data
+     * store to attempt evaluation
+     * @param sorting  Sorting which can be pushed down to the data store.
+     * @param pagination  Pagination which can be pushed down to the data store.
+     * @param scope  Contains request level metadata.
+     *
+     * @return a optimized query consumed by the data store
+     */
+    Query buildQuery(
+            Class<?> entityClass,
+            Optional<WhereFilterExpression> whereFilter,
+            Optional<HavingFilterExpression> havingFilter,
+            Optional<Sorting> sorting,
+            Optional<Pagination> pagination,
+            RequestScope scope
+    );
+
+    /**
+     * Given a {@link FilterExpression}, this method gathers all the WHERE filter clauses together into a single WHERE
+     * clause.
+     *
+     * @return a filter expression that can be executed as a "WHERE"query constraint
+     */
+    Optional<WhereFilterExpression> extractWhereExpression(Optional<FilterExpression> filterExpression);
```

Comment:
I think this should be part of the base class for the aggregation data store rather than part of the QueryEngine.  QueryEngine should have a very simple contract.

Reply:
Ok. I will only contract query building and query execution then